src/core/config/config.rs

```diff
@@ -510,17 +510,47 @@ impl Config {
         types
     }
 
-    /// Returns a list of all the types that are used as interface
-    pub fn interface_types(&self) -> HashSet<String> {
-        let mut types = HashSet::new();
+    pub fn interfaces_types_map(&self) -> BTreeMap<String, BTreeSet<String>> {
+        let mut interfaces_types: BTreeMap<String, BTreeSet<String>> = BTreeMap::new();
+
+        for (type_name, type_definition) in self.types.iter() {
+            for implement_name in type_definition.implements.clone() {
+                interfaces_types
+                    .entry(implement_name)
+                    .or_default()
+                    .insert(type_name.clone());
+            }
+        }
 
-        for ty in self.types.values() {
-            for interface in ty.implements.iter() {
-                types.insert(interface.clone());
+        fn recursive_interface_type_merging(
+            types_set: &BTreeSet<String>,
+            interfaces_types: &BTreeMap<String, BTreeSet<String>>,
+        ) -> BTreeSet<String> {
+            let mut types_set_local = BTreeSet::new();
+
+            for type_name in types_set.iter() {
+                match interfaces_types.get(type_name) {
+                    Some(types_set_inner) => {
+                        let types_set_inner =
+                            recursive_interface_type_merging(types_set_inner, interfaces_types);
+                        types_set_local = types_set_local.merge_right(types_set_inner);
+                    }
+                    None => {
+                        types_set_local.insert(type_name.to_string());
+                    }
+                }
             }
```

Comment:
small note about possible optimization here:
we recalculate the types every time even if we have already calculated some subtrees

for this example:
```graphql
interface T1 {}
interface T2 implements T2 {}
interface T3 implements T1 & T2 {}
type A implements T1 & T2 & T3 {}
```

when executing T1 we will calculate that A belongs to T2 through T3, but when we start processing T2 we will rerun all the checks through T3 again.
Hope it's not a big deal =)